torch/distributed/_sharded_tensor/api.py

```diff
@@ -237,6 +253,83 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._barrier(rpc_workers)
 
+    @classmethod
+    def _init_from_local_shards(
+        cls,
+        local_shards: List[Shard],
+        sharded_tensor_metadata: ShardedTensorMetadata,
+        process_group=None
+    ):
+        shards_metadata = sharded_tensor_metadata.shards_metadata
+
+        if len(shards_metadata) == 0:
+            raise ValueError("shards_metadata must not be empty!")
+
+        sharded_tensor = cls.__new__(cls)
+
+        # prepare initialization
+        sharded_tensor._prepare_init(process_group=process_group)
+
+        sharded_tensor._metadata = sharded_tensor_metadata
+
+        # no sharding spec for sharded tensors that initialized
+        # from this API.
+        sharded_tensor._sharding_spec = None
+
+        current_rank = dist.get_rank(sharded_tensor._process_group)
+
+        local_shard_metadatas = []
+
+        # collect local shard metadatas from the global sharded_tensor_metadata
+        for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_metadata.placement)
+
+            if current_rank == rank:
+                local_shard_metadatas.append(shard_metadata)
+
+        if len(local_shards) != len(local_shard_metadatas):
+            raise RuntimeError(
+                f'Number of local shards ({len(local_shards)}) does not match number of local '
+                f'shards metadata in sharded_tensor_metadata ({len(local_shard_metadatas)}) '
+                f'on rank ({current_rank}) '
+            )
+
+        for shard in local_shards:
+            shard_meta = shard.metadata
+            local_shard_tensor = shard.tensor
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_meta.placement)
+
+            # validate if shard_meta in the metadatas collected from sharded_tensor_metadata
+            assert shard_meta in local_shard_metadatas, \
+                "local shard metadata not in sharded_tensor_metadata!"
+
+            if shard_meta.shard_lengths != list(local_shard_tensor.size()):
+                raise ValueError(
+                    f'Local shard tensor is incompatible with local ShardMetadata! '
+                    f'local shard tensor size: {local_shard_tensor.size()}, '
+                    f'local ShardMetadata shard lengths: {shard_meta.shard_lengths}'
+                )
+
+            if str(local_shard_tensor.device) != local_device:
```

Comment:
We should also verify that the `dtype`, `layout`, `requires_grad` etc. of local_shards is same as what is specified in ShardedTensorMetadata.

Can we also add unit tests where this contract is not respected and we throw an appropriate error?

Reply:
good point, just added